### DISCUSSION TITLE
Help! How to use dtfold ?

### DISCUSSION THREAD

**Discussion by pbreuer** [clash-lang/clash-compiler#2400] [Q&A] [answered]
Sorry folks, but I can't work it out from the example+description in the manual at https://hackage.haskell.org/package/clash-prelude-1.6.4/docs/Clash-Sized-Vector.html#v:dtfold . Can anyone help?

I am trying to implement a vector mux:

    muxN  :: ( HiddenClockResetEnable dom, KnownNat n) 
         => Signal dom (Index (2^n)) -> Vec (2^n) (Signal dom x) -> Signal dom x

for which the obvious, clean, sane simple implementation unfortunately synthesizes using many tens of thousands of logic units, say 60 thousands.  My vector is 256 wide and each wire ("x") is some hundreds of bits wide. Yes, that sane implementation is

     muxN is sv = fmap f (bundle(bundle sv,is))
                     where f (v,i) = v !! i

This is a candidate to get the complexity down from 2^n to n log2 n in the classical way by  recursive halving. The basic idea is ... do v !! i   using the top bit of index i to select the first or the second of the two halves v1,v2 of v=v1++v2, and then recursively use the remaining bits of  index i to select from inside the half-vector. Perfect for dtfold?

Everything I try elicits incomprehensible (to me) type complaints about needing KnownNat l !-(.

I must use vector v :: Vec (2^n) x to construct a function f :: Index (2^n) -> x

So for dtfold I need to specify how, given two such functions f1,f2 constructed from the two halves v1,v2 with v1++v2=v, I make function f for v:

    f = \i -> if msb i == 0 then f1 i' else f2 i'   where i' = lowbits i

(give me the lowbits function, say lowbits i = unpack i' where (_::BitVector 1,i') = split i)

This is the  combinator  among the parameters to the dtfold function, though one needs to also say what level "n" it is at. Call it

    combi :: KnownNat n => SNat n -> (Index (2^n) -> x) -> (Index (2^n) -> x) -> (Index (2^(n+1)) -> x)
    combi n f1 f2 = \i -> if msb i == 0 then f1 i' else f2 i'   where i' = lowbits i

The other argument to dtfold is the base of the recursion/induction, applied to vectors length 1, i.e. the elements of the original vector v. Like this:

     bse :: x -> (Index (2^0) -> x)
     bse x = \i -> x

(that index i is in the Index 1 type, which makes it necessarily 0, so we know what  it is and ignore it).
    
These components plus some mysterious type formula that I will call foo for now are put together by dtfold :

     dtfold (Proxy @foo) bse combi :: Vec (2^n) x -> (Index (2^n) -> x)

so I should get

    (!!!) :: KnownNat n => Vec (2^n) x -> Index (2^n) -> x
    v !!! i = dtfold (Proxy @foo) bse combi v i

and as far as I can make out from the manual that foo thing is given by

    data IIndex (x::Type) (f :: TyFun Nat Type) :: Type
    type instance Apply (IIndex x) n = Index (2^n) -> x

and I want (???) Proxy @(IIndex x) where I have written Proxy @foo.

The above definition of (!!!) gives the single error

    * Could not deduce (KnownNat l) arising from a use of `combi'
      from the context: KnownNat n
        bound by the type signature for:
                   (!!!) :: forall (n :: Nat) x.
                            KnownNat n =>
                            Vec (2 ^ n) x -> Index (2 ^ n) -> x
        at KPU/TACache.hs:1006:1-78
      Possible fix:
        add (KnownNat l) to the context of
          a type expected by the context:
            forall (l :: Nat).
            SNat l -> (IIndex x @@ l) -> (IIndex x @@ l) -> IIndex x @@ (l + 1)
     In the third argument of `dtfold', namely `combi'
      In the expression: dtfold (Proxy @(IIndex x)) bse combi
      In an equation for `!!!':
          (!!!) = dtfold (Proxy @(IIndex x)) bse *combi*

There is no "l" (letter "ell") anywhere in my definitions. I have declared KnownNat n of every n::Nat that I have introduced in these definitions. What is going on? Is this my mistake or a mistake in dtfold somewhere? 

The dtfold function wants its argument "combi" to have a universally quantified type that takes a natural as first argument, and combi is defined so that is so. What am I supposed to do here?

Thanks for any light in the darkness.

PTB

**Comment by lmbollen** [comment]
Thanks for opening this discussion!

I opened the documentation for `dtfold` and the first thing I spot is a type variable `l` in the third argument, that's probably what the compiler is referring to. 

Could you try explicitly applying `n` to `combi` by changing `(!!!)` to:
```haskell
(!!!) :: forall n . KnownNat n => Vec (2^n) x -> Index (2^n) -> x
v !!! i = dtfold (Proxy @foo) bse (combi @n)  v i
```

**Reply by pbreuer** [reply]
The result of an @n on combi  is a slightly different you-can't-do-that-there error :-(

    * Couldn't match type `l' with `n'
      Expected: SNat l
                -> (IIndex x @@ l) -> (IIndex x @@ l) -> IIndex x @@ (l + 1)
        Actual: SNat n
                -> (Index (2 ^ n) -> x)
                -> (Index (2 ^ n) -> x)
                -> Index (2 ^ (n + 1))
                -> x
      `l' is a rigid type variable bound by
        a type expected by the context:
          forall (l :: Nat).
          SNat l -> (IIndex x @@ l) -> (IIndex x @@ l) -> IIndex x @@ (l + 1)
        at KPU/TACache.hs:1010:40-49
      `n' is a rigid type variable bound by
        the type signature for:
          (!!!) :: forall (n :: Nat) x.
                   KnownNat n =>
                   Vec (2 ^ n) x -> Index (2 ^ n) -> x
        at KPU/TACache.hs:1009:1-78
    * In the third argument of `dtfold', namely `(combi @n)'
      In the expression: dtfold (Proxy @(IIndex x)) bse (combi @n)
      In an equation for `!!!':
          (!!!) = dtfold (Proxy @(IIndex x)) bse **(combi @n)**
    * Relevant bindings include
        (!!!) :: Vec (2 ^ n) x -> Index (2 ^ n) -> x
        

To me, it looks like the combi argument to dtfold needs to have a type that is valid for all possible naturlals l (or n), without the condition that is a KnownNat.

But I can't write such a thing because split, splice and anything else I try using in combi (for lowbits) end up requiring a KnownNat.

Any clue there? Is  there some deeper interface I can get to somehow that will work to define lowbits without requiring the KnownNat l (or n).

 On the face of it split would do because it only requires KnownNat of the second result, and I can use BitVector 1 for that in lowbits i = unpack i1 where (i1,i2::BitVector 1) = split (shiftL (pack i) 1)

    split  :: (BitPack a, KnownNat n, BitSize a ~ (m + n)) => a -> (BitVector m, BitVector n)

But no, it lies and says KnownNat is required, leading to the complaint for (!!!) that KnownNat l is needed for the combi arg to dtfold.

**Reply by leonschoorl** [reply]
It doesn't lie.

The fact that is complains about a missing `KnownNat n` is caused by the that you're calling `split` (via `lowbits` on a `Index (2^(n+1)`.
The `split` needs a `BitPack` instance on its first argument.
And the BitPack instances for "sized" types; like BitVector, Signed, Unsigned, Index, etc; are defined with the KnownNat constraint.
If you look at the [documentation for BitPack](https://hackage.haskell.org/package/clash-prelude-1.6.4/docs/Clash-Class-BitPack.html#t:BitPack) and scroll down to "Instances", you can see the the instance for `BitPack (Index n)` is defined with the constraint `(KnownNat n, 1 <= n)`.
So that extra KnownNat is sort of hiding inside the `BitPack a`.

Luckily[^1] in this case we have an `SNat`, and an SNat [contains an KnownNat constraint](https://hackage.haskell.org/package/clash-prelude-1.6.4/docs/Clash-Promoted-Nat.html#t:SNat)
All we have to do to get it out and make it accessible to typechecker is pattern match on the `SNat` constructor.
Full example:
```haskell
module TestDtfold where
import Clash.Prelude
import Data.Singletons (Apply, TyFun, Proxy(Proxy))

topEntity :: Vec 8 Int -> Index 8 -> Int
topEntity = (!!!)

lowbits :: (BitPack a, BitPack b, BitSize a ~ BitSize b + 1) => a -> b
lowbits i = unpack i' where (_::BitVector 1,i') = split i

combi :: SNat j -> (Index (2^j) -> x) -> (Index (2^j) -> x) -> (Index (2^(j+1)) -> x) 
combi SNat f1 f2
  = \i -> let i' = lowbits i in if msb i == 0 then f1 i' else f2 i'

bse :: x -> (Index (2^0) -> x)
bse x = \i -> x

data IIndex (x::Type) (f :: TyFun Nat Type) :: Type
type instance Apply (IIndex x) n = Index (2^n) -> x

(!!!) :: forall n x . KnownNat n => Vec (2^n) x -> Index (2^n) -> x
v !!! i = dtfold (Proxy @(IIndex x)) bse combi v i
```

[^1]: actually it's not luck, `dtfold`'s third argument is of the type `forall l. SNat l -> ...` precisely so we can do these kinds of things.

**Reply by leonschoorl** [reply]
What [christiaanb said](https://github.com/clash-lang/clash-compiler/discussions/2400#discussioncomment-4634972)

**Comment by rowanG077** [comment]
Note that I would expect any and all synthesis tools to already do this by default. I tested with yosys for ECP5 and it does use a tree like structure. Not that it's not possible to do it structurally using `dtfold` but indexing is such a low-level operation that I doubt it wouldn't be optimized like this.

For this `topEntity` (ignore the `unpack` it's only to make the diagram prettier):

```
topEntity :: Index 8 -> BitVector 64 -> BitVector 8
topEntity is vs = unpack vs !! is
```

it synthesizes to this circuit for target ECP5 (For some reason it renders incorrectly when opening the link directly. Please save the svg and view it that way) :

![download](https://user-images.githubusercontent.com/7439756/211304290-027fdc9f-d388-4447-bc49-c32660987dc0.svg)

**Comment by pbreuer** [comment]
Thanks! (nice diagram!) I'm running yosys for xilinx. The vector is 256 wide, the vector elements are 633 bits wide.  Used in the muxN of a 256-vector of signals, fmapping the (!!) timewise along the signals, as exhibited in the post at head of thread, that is  66231  LCs. Can you confirm?

    === Synth_muxN ===

     Number of wires:              51141
     Number of wire bits:         447167
     Number of public wires:         260
     Number of public wire bits:  324740
     Number of memories:               0
     Number of memory bits:            0
     Number of processes:              0
     Number of cells:              91431
     LUT1                           90
     LUT2                          644
     LUT3                         2355
     LUT4                        22629
     LUT5                         1061
     LUT6                        40186
     MUXF7                       24080
     MUXF8                         386

     Estimated number of LCs:      66231

(I'd still like to know how to use dtfold! It may be useful! For one thing I could save yosys optimization time even if it does manage to come out with that tree structure -- which is a feat of improbable reverse engineering that i am utterly amazed by)

Ironically, I like this message from yosys:

    4.25.3. Executing OPT_MUXTREE pass (detect dead branches in mux trees).
    Running muxtree optimizer on module \Mux..
    Creating internal representation of mux trees.
    No muxes found in this module.

**Reply by rowanG077** [reply]
That circuit is too large to visualize. `netlistsvg` just balks so I can't confirm it definitely does it. But this `topEntity`:

```
type Elem = BitVector 16
type Depth = 5

topEntity :: Signal System (Index (2^Depth)) -> Vec (2^Depth) (Signal System Elem) -> Signal System Elem
topEntity is sv = fmap f (bundle(bundle sv,is))
  where f (v,i) = v !! i
```

Definitely synthesizes to a tree structure using `synth_xilinx`. With the following resource usage:

```
=== topEntity ===

   Number of wires:                215
   Number of wire bits:           1912
   Number of public wires:          67
   Number of public wire bits:    1557
   Number of memories:               0
   Number of memory bits:            0
   Number of processes:              0
   Number of cells:                792
     IBUF                          517
     LUT2                            9
     LUT3                           17
     LUT4                           49
     LUT6                          134
     MUXF7                          50
     OBUF                           16

   Estimated number of LCs:        200
```

With the size you used `yosys` `techmap` takes ages so I killed it.

**Reply by rowanG077** [reply]
Synthesizing with a few different configs

```
Depth   Elem     LC
8       128      13388
8       64       6599
8       32       3333
7       128      6897
7       64       3471
7       32       1701
6       128      3066
6       64       1504
6       32       751
```

This seems to be best it could theoretically do? If you increase the depth by 1 you double the logic cells since you now have two subtrees of the same size as the first one. If you double the element size you also double the logic cells since you need to select that many bits.

So Logic cell usage is `2^Depth * ElemSize` and circuit depth is just `Depth`.

**Reply by pbreuer** [reply]
@rowanG077 yes, thank you, that is very thorough of you. Yes, it also takes several hours (3h) here to synthesize with those 600+ bits wide wires and 256-dim vector,  on this AMD 2.2GHz multicore machine. taking  4.5GB ram. You are correct ... the number of LCs is proportional to the width of the wires. I have split width wise and run two halves in parallel for the same behavior and about the same complexity, other than some duplicated circuit overhead.

But no, I don't see that the number of logical primitives should be exponential in the depth. Isn't the depth the number of bits in the indexing, i.e. the exponent in the dimension of the vector of streams. So  2^depth is the number of streams N and you are saying N * W primitive logical components where W is the width of the streams. The classical number from recursive halving is  N * depth * W.

From your table: 

    D   N   W          n      n/(D*N*W)
    8  256  128      13388     0.05 
    8  256  64       6599      0.05
    8  256  32       3333      0.05
    7  128  128      6897      0.06 
    7  128  64       3471      0.06 
    7  128  32       1701      0.06 
    6  64   128      3066      0.06
    6  64   64       1504      0.06
    6  64   32       751       0.06 

My datum was:

    8  256  633      66231     0.05

**Reply by rowanG077** [reply]
@pbreuer 

You can approximate the structure with a perfect binary tree. Each node is logic cell. A perfect binary tree of depth `d` has `2^(d+1) – 1` nodes.  I would expect the number of LC is linked to the number of nodes in the tree. 

```
D   N   W          n      n/(D*N*W)  n / (2^D * W)
8  256  128      13388     0.05       0.41
8  256  64       6599      0.05       0.40
8  256  32       3333      0.05       0.41
7  128  128      6897      0.06       0.42
7  128  64       3471      0.06       0.42
7  128  32       1701      0.06       0.41
6  64   128      3066      0.06       0.37
6  64   64       1504      0.06       0.37
6  64   32       751       0.06       0.37
```

and

```
8  256  633      66231     0.05       0.41
```

**Reply by pbreuer** [reply]
The structure also breaks out each bit of the index to drive in a different layer of the tree. That ends up as an extra multiplier in the complexity by the number of bits / aka depth of the tree.

 This is "intuitively" - I haven't tried drawing out the diagram and actually checking the intuition because it is so much part of learned and practiced lore that the complexity is n * log2 n! Maybe you're right! 

That the "constant of proportionality"  is increasing with scale in your table argues against, however. Capital overheads should diminish in proportion with scale.  Increasing points to a missed contributing factor.

Management costs :-!

PS. I generated another line of that table, for greater depth D:

    D  N    W  n      n/DNW   n/NW
    10 1024 32 14286  .04     .434

The last column number is still rising, so does not look to be bounding. The penultimate column number is still falling, so is apparently bounding. Falling is what I expect as capital overheads become proportionately less with increasing scale.

**Comment by christiaanb** [comment, marked as answer]
Try this:
```
combi :: SNat n -> (Index (2^n) -> x) -> (Index (2^n) -> x) -> (Index (2^(n+1)) -> x)
combi SNat f1 f2 = \i -> if msb i == 0 then f1 i' else f2 i'   where i' = lowbits i
```
Pattern matching on the `SNat` (of type `SNat n`) introduces the `KnownNat n` constraint for anything in scope of the pattern match. As a result, you no longer need the `KnownNat n` constraint in the type signature of `combi`, since anything inside the definition of `combi` will now get the `KnownNat n` out of the `SNat`.

**Comment by pbreuer** [comment]
@christiaanb Yes, that works!  No "KnownNat n" in the definition of the combinator but pattern match on SNat as first argument instead. That seems like a technique!
I'll give results presently.
Marked as best answer because it solves the literal problem (any answer from Christiaan is almost cheating!) but  I am very interested in more observations on this question. I am surprised in one sense that clash does not force a binary tree implementation for !! (it doesn't?) but not surprised  in the sense that optimization is  the real business of a back end and it is good software engineering not to try for it. Still, the binary tree would always be better, I think. Even if only half the tree (+1) is needed, that will be n log2 n complexity instead of 2^(n-1) complexity, and that is the least advantageous case.

**Comment by pbreuer** [comment]
With dtfold implementing !!, the results after synthesis with yosys for xilinx are the same as when using the standard !!. So yosys seems to automatically make a tree structure for xilinx too ... 

(256-way 633-bit wide multiplexer)

    === Synth_muxN ===

     Number of wires:              55257
     Number of wire bits:         1954281
     Number of public wires:        1667
     Number of public wire bits:  1826863
     Number of memories:               0
     Number of memory bits:            0
     Number of processes:              0
     Number of cells:             257371
     IBUF                       162315
     LUT1                          164
     LUT2                         1124
     LUT3                         2712
     LUT4                        23568
     LUT5                          838
     LUT6                        39898
     MUXF7                       25324
     MUXF8                         794
     OBUF                          634

     Estimated number of LCs:      67016


Does anyone have any better ideas? This is very costly for such trivial functionality. Is it possible to use three-valued logic somehow? I could power down all but one of the incoming vector of streams very cheaply with a demuxer driven from the stream of incoming indices, and just join all the wires without any gating ....

Thanks for all the inputs!

**Reply by leonschoorl** [reply]
I think this is why `!!` in clash-prelude isn't "optimised" in this way.
Because it generally synthesizes down to the same hardware anyway.
And the standard `!!` will probably be faster when running simulations in Haskell.

As for the three-valued logic idea, I wouldn't expect any synthesis packages would let you do that on an FPGA.

**Reply by rowanG077** [reply]
@pbreuer @leonschoorl Is a sequential implementation not an option? If the throughput and latency drop is acceptable.

An implementation could be to flip around a single sub-vector according the relevant bit in the index every cycle. Ensuring that after `n` cycles for an `Index 2^n` the required element is in the head of the vector.

There is also an additional cost to be paid in multiplexers, registers and other logic. But I don't think it's as high as the combinatorial implementation.

I'm not 100% sure this will actually synthesize to smaller hardware but it could be worth a try.

**Reply by pbreuer** [reply]
The problem is the sheer number of LCs required by this really quite trivial component, which squeezes what is available for everything else that has to go in on an FPGA.

The suggestion is to swap time for space via a multi-cycle implementation?

Can't do :-(. This is  reading a line of a 256-line fast cache. Of course, if the platform provides caches, one could use THAT. One would need a cache primitive in clash.

**Reply by rowanG077** [reply]
@pbreuer You should use RAM for cache. BlockRAM is preferable if you can deal with the latency of 1 cycle. 

This:

```
type N = 8
type W = 663

topEntity
  :: Clock System
  -> Reset System
  -> Enable System
  -> Signal System (Index (2^N))
  -> Signal System (Maybe (Index (2^N), BitVector W))
  -> Signal System (BitVector W)
topEntity clk rst en = blockRamU clk rst en NoClearOnReset (SNat @(2^N)) (const undefined)
```

Synthesizes almost instantly with this report:

```
=== topEntity ===

   Number of wires:                 18
   Number of wire bits:           4064
   Number of public wires:          10
   Number of public wire bits:    2689
   Number of memories:               0
   Number of memory bits:            0
   Number of processes:              0
   Number of cells:               1367
     BUFG                            1
     IBUF                          683
     LUT2                            1
     OBUF                          663
     RAMB18E1                       19

   Estimated number of LCs:          1
```

Even `asyncRAM` works and that doesn't have latency. 

```
type N = 8
type W = 663

topEntity
  :: Clock System
  -> Enable System
  -> Signal System (Index (2^N))
  -> Signal System (Maybe (Index (2^N), BitVector W))
  -> Signal System (BitVector W)
topEntity clk en = asyncRam clk clk en (SNat @(2^N))
```
```
=== topEntity ===

   Number of wires:               1340
   Number of wire bits:           7355
   Number of public wires:           9
   Number of public wire bits:    2696
   Number of memories:               0
   Number of memory bits:            0
   Number of processes:              0
   Number of cells:               2897
     BUFG                            1
     IBUF                          682
     LUT4                            4
     LUT6                          663
     OBUF                          663
     RAM64M                        884

   Estimated number of LCs:        667
```

**Reply by pbreuer** [reply]
That's really helpful!  I thought blockram was for sdram (about 5ns cycles), not cache (about 1ns cycles).  I was using regEn at the heart of it, plus logic.  Do platforms offer 1ns blockram?

**Reply by rowanG077** [reply]
Most (I think all) FPGAs offer blockram. I don't think there are FPGAs that can reasonably clock to Ghz level speeds let alone run blockRAM at that speed. And if they exist they are far above my paygrade.

Others here might know more about such high-end parts.

**Reply by pbreuer** [reply]
That is very, very helpful indeed! I thought blockram was for things like sdram (5ns cycle), not cache/registers (1ns cycles).

A problem with blockRams is that it takes 256 cycles to clear via reset (internal sequence, obviously) and that's not acceptable here. Otherwise OK. AsyncRam doesn't seem to reset at all. It would take  256 cycles to clear it externally.

What's the deal with reset and  ram?

(I like the trueDualPortBlockRam! More like that please! I need one write and three read ports, so I would be tripling up plain blockRams in parallel to read from three different addresses simultaneously.)

**Reply by pbreuer** [reply]
Thanks again! Ahem, the technique is to obtain parts that satisfy the design, not the other way round! (Some call this specification followed by engineering, others call it  funding by what you can promise, not what you can already show). Top end xilinx or Intel fpga indeed, heading for asic.

Don't (physical) sdrams have a single cycle reset line threaded through the whole thing? It just has to briefly short a flipflop or a transistor or something ... Well, I'll add that to the specification! 

A 256 cycle delay on reset is livable for now. I could even swap out one blockRam for another to simulate one clearable device.

**Reply by pbreuer** [reply]
I am told "Quite a few of the FPGA designs come with significant amounts of onboard cache as they are used to develop specialized accelerators.". So it would be possible to write primitives that map to them., either in clash or yosys or both. I know how to do neither! (yes, I have seen the tutorial :-).

a simple pull-through cache would have type:

    Signal dom  (Maybe (a,Bool))         --  incoming read/invalidate cmd to front of cache with address
    -> Signal dom  (Maybe(a,x))               -- incoming address+data writes from behind cache
    -> ( Signal dom (Maybe(a,Maybe x))  -- outgoing address+data response from front of cache
     , Signal dom (Maybe a)                 -- outgoing query to memory back of cache with address
     )

Plus a reset line of course! And some configuration parameters like n=log2 #lines (each an x). Internally n bits of the address will designate the line and the rest of the address bits will form a tag for each line, so one will need at least 2^n * (|x| + |a| - n) bits of internal  space. There might be some extra bits for dirty/uptodate etc. flags so one can invalidate a line or age it, etc.

Another parameter might specify if it is one-way, two-way, 4-way associative etc. Two way means doubled storage inside, with two lines stored for each |a| - n bits of address and the older of the two being replaced on write.

Another parameter might specify which bits of address are indexing. Usually middle bits.

I'll look out specs for a cache that can be written from the front too (as well as from the back; the above can only be asked for read from the front). Or tell me if you know! I just make this stuff up ...

Are there any plans to target mapping to these components?

**Reply by rowanG077** [reply]
BlockRAM is the cache they mention. You will have to build your specific cache out of the blockRAM primitive.

**Reply by pbreuer** [reply]
Yes, they seem to be talking about "ultraram" (too) but as "replacement for external sdram" so one could deduce the working cycle will be roughly commensurate with sdram and some people have said it is as fast as 3 ns to me. "The fastest will be block ram (either Xilinx or Intel). It will be more than 1 cycle (except as lower speed for the most part). However, you can pipeline it such that you achieve it. Some FPGAs have embedded DRAM for this L1 type use, much faster than DRAM external on the board ..." - make what you will of that.

There are fast caches on board these high-end FPGA platforms and the promotional literature seems to highlight them but I can't see how they are directly accessible to the FPGA design. Still, if they cache memory calls through the platform, then there may be no need to cache data to/from memory at all within the FPGA design. One can remove those caches from the design.

In the meantime I have checked and one can make up for the lack of a 1-cycle reset line on blockRamU (and friends) in Clash by adding a filtering bitmap that records dirtied (i.e. written to) elements in the ram. Reading from an  element said by the bitmap to be undirtied now returns 0 without consulting the ram. To effectively reset, just clear the filtering bitmap so every element is seen as undirtied again and 0 is returned for it.

**Reply by rowanG077** [reply]
ultraRAM is xilinx specific. UltraRAM is basically blockRAM + some restrictions that standard Xilinx blockRAM doesn't have. I expect ultraRAM be inferred from the Clash blockRAM primitive. 

https://docs.xilinx.com/r/en-US/am007-versal-memory/Block-RAM-and-UltraRAM-Differences

blockRAM and ultraRAM are all documented by Xilinx. I'm not sure what you mean that they aren't accessible.

**Reply by pbreuer** [reply]
(My thanks to everyone who has participated here, by the way).

However .... it appears not quite so easy to use blockRAM for cache. This cache has two write ports and two read ports and blockRAM has one write port and one read port. Two read ports can be simulated by running two blockRAMs in parallel, writing the same thing to both of them. But, but, but ...

... I can't write two times at once. OK, you say, so use trueDualPortBlockRAM instead. Yes, that has two  ports  but they are not duplex, so I cannot read at the same time as two writes are happening.

Alright, that's only going to cause the cache to drop some read requests, so the requester will go to memory instead for those. And provided two  writes are not happening at once I can read at two different places at once, by the same trick of putting two trueDualPortBlockRAMs in parallel.

But then the read response is delayed one clock in block rams in general. That is, with a register implementation for caches, a read request gets immediately the value from one cycle ago, before the current write happened.  Using a block ram a read request will produce nothing immediately, and one cycle later the value comes out, and I can't tell from the documentation if that will reflect the current write or not. Hopefully, being late, the response will at least take it into account.

I might be able to resolve most of these difficulties with external logic, but not all, I (don't) think.
